### ISSUE TITLE
terminate called after throwing an instance of 'shark::Exception'   what():  [import_csv_reader_points] problems parsing file (3)

### ISSUE THREAD

**Issue by amaad** [Shark-ML/Shark#195]
i am facing this error can anyone guide me here is my code . Error in this line Instances trainInstances(trainFiles[t* 10 + i]);

main5.cpp
<
#include <iostream>
#include <vector>
#include <fstream>
#include <string>

#include <shark/ObjectiveFunctions/CrossValidationError.h>
#include <shark/Algorithms/DirectSearch/GridSearch.h>
#include <shark/Algorithms/JaakkolaHeuristic.h>

#include <shark/Algorithms/Trainers/CSvmTrainer.h> // the C-SVM trainer
#include <shark/Models/Kernels/GaussianRbfKernel.h> //the used kernel for the SVM
#include <shark/ObjectiveFunctions/Loss/ZeroOneLoss.h> //used for evaluation of the classifier
#include <shark/Data/DataDistribution.h> //includes small toy distributions

#include <shark/Algorithms/Trainers/NormalizeComponentsUnitVariance.h>
#include <shark/Algorithms/Trainers/NormalizeComponentsUnitInterval.h>
#include <shark/Algorithms/Trainers/NormalizeComponentsWhitening.h>
#include <shark/Algorithms/Trainers/NormalizeComponentsZCA.h>
#include <shark/Algorithms/Trainers/NormalizeKernelUnitVariance.h>
#include <shark/Models/Normalizer.h>
#include <shark/Models/LinearModel.h>

#include "Instances.h"
#include "SVGPDTM.h"

using namespace std;

int main(int argc, char** argv)
{
	vector<string> trainFiles;
	vector<string> testFiles;

        std::ifstream file(argv[1]);

	std::string str;
	while (std::getline(file, str))
	{
		trainFiles.push_back(str);
	}

	file.close();

	file.open(argv[2]);
	while (std::getline(file, str))
	{
		testFiles.push_back(str);
	}

	file.close();

	int TP = 0; int FP = 0;
	int FN = 0;	int TN = 0;
	int nSV = 0;
	
	
	double C = atof(argv[3]);
//        std::cout << C << std::endl;
	double gamma = atof(argv[4]);
	int population = atoi(argv[5]);
	int gen = atoi(argv[6]);
	int setCount = (atoi(argv[7]));

	int negWeight = atoi(argv[8]);
	int posWeight = atoi(argv[9]);
	int compareSVM = atoi(argv[10]);
	string outputResult = argv[11];
	
	bool removeMean = true;

	//Normalization components
	//Normalizer<RealVector> normalizer;
	//NormalizeComponentsUnitVariance<RealVector> normalizingTrainer(removeMean);
	
	//Metric Attributes Blue Print
	typedef struct _MetricAttribute
	{
		std::string cvType;
		double Accuracy = 0.0;
		double GMean = 0.0;
		double nSV = 99999999;
		double C = 0.0;
		double gamma = 0.0;

		int TP; int FP;
		int FN; int TN;
	} 
        MetricAttribute;

	const std::string cvString[] = { "dbscv", "dobscv", "scv" };
	MetricAttribute bestFromOriginal[3];

	std::vector<MetricAttribute> svmPerformance;
	std::vector<MetricAttribute> svgpmPerformance;

	for (int g = -4; g < 7; ++g)
	{
		double gamma = std::pow(2, g);
		for (int c = -2; c < 8; ++c)
		{
			double C = std::pow(2, c);
			MetricAttribute ma;
			ma.C = C;
			ma.gamma = gamma;

			svmPerformance.push_back(ma);
			svgpmPerformance.push_back(ma);
		}
	}
        
//	svmPerformance.shrink_to_fit();
//	svgpmPerformance.shrink_to_fit();

	//SVM
	std::vector<std::vector<MetricAttribute>> listPerformance;
	for (int t = 0; t < 3; ++t)
	{
	    if (compareSVM == 0)
		break;
#pragma omp parallel for private(TP,FP,FN,TN,nSV) 
		for (int p = 0; p < svmPerformance.size(); ++p)
		{

			nSV = 0;
			TP = 0; FP = 0;
			FN = 0;	TN = 0;
			for (int i = 0; i < 10; ++i)
			{
                            cout << trainFiles[t* 10 + i] <<endl;
                          
                                Instances trainInstances(trainFiles[t* 10 + i]);
				Instances testInstances(testFiles[t* 10 + i]);
                            
				Normalizer<RealVector> normalizer;
				////LinearModel<RealVector> normalizer;
				////NormalizeComponentsUnitVariance<RealVector> normalizingTrainer(removeMean);
				NormalizeComponentsUnitInterval<RealVector> normalizingTrainer;
				////NormalizeComponentsWhitening normalizingTrainer;
				normalizingTrainer.train(normalizer, trainInstances.GetLabeledData().inputs());
				LabeledData<RealVector, unsigned int> normalizedTrainData = transformInputs(trainInstances.GetLabeledData(), normalizer);
				LabeledData<RealVector, unsigned int> normalizedTestData = transformInputs(testInstances.GetLabeledData(), normalizer);
                                   
				Instances normalizedTrainInstances(normalizedTrainData);
				Instances normalizedTestInstances(normalizedTestData);
				
				shark::GaussianRbfKernel<RealVector> rbffst(svmPerformance[p].gamma);
				shark::SquaredHingeCSvmTrainer<RealVector> shcsvmfst(&rbffst, svmPerformance[p].C * negWeight, svmPerformance[p].C * posWeight);
				shark::KernelClassifier<RealVector> kcfst;
				////shcsvmfst.s2do() = true;
				shcsvmfst.train(kcfst, normalizedTrainData);
				nSV += kcfst.decisionFunction().alpha().size1();
				Data <unsigned int> rst = kcfst(normalizedTestData.inputs());

				for (int j = 0; j < rst.numberOfElements(); ++j)
				{
					int original = normalizedTestData.element(j).label == 0 ? 0 : 1;
					if (original == 0 && rst.element(j) == 0)
					{
						++TN;
					}
					else if (original == 0 && rst.element(j) == 1)
					{
						++FP; //if actual is negative, but output is positive, then it is false positive
					}
					else if (original == 1 && rst.element(j) == 0)
					{
						++FN; //if actual is positive, but output is negative, then it is false negative
					}
					else if (original == 1 && rst.element(j) == 1)
					{
						++TP;
					}
				}
			}	

			double Specificity = (double)TN / (double)(TN + FP);
			double Sensitivity = (double)TP / (double)(TP + FN);
			double Accuracy = (double)(TP + TN) / (double)(TP + TN + FP + FN);
			double GMean = std::sqrt(Specificity * Sensitivity);

			svmPerformance[p].Accuracy = Accuracy;
			svmPerformance[p].GMean = GMean;
			svmPerformance[p].TP = TP;
			svmPerformance[p].FP = FP;
			svmPerformance[p].FN = FN;
			svmPerformance[p].TN = TN;
			svmPerformance[p].nSV = nSV;
			svmPerformance[p].cvType = cvString[t];
			if (bestFromOriginal[t].GMean < svmPerformance[p].GMean)
				bestFromOriginal[t] = svmPerformance[p];
		}
		listPerformance.push_back(svmPerformance);
	}

	if (compareSVM != 0)
	{
		ofstream fout("data/resultSVM.txt");
		fout << "CVType\tC\tgamma\tTP\tFP\tFN\tTN\tAccuracy\tGmean\tnSV\n";
		for (int p = 0; p < listPerformance.size(); ++p)
		{
			for (int q = 0; q < listPerformance[p].size(); ++q)
			{
				fout << listPerformance[p].at(q).cvType << "\t" << listPerformance[p].at(q).C << "\t" << listPerformance[p].at(q).gamma << "\t" << listPerformance[p].at(q).TP << "\t" << listPerformance[p].at(q).FP
					<< "\t" << listPerformance[p].at(q).FN << "\t" << listPerformance[p].at(q).TN << "\t" << listPerformance[p].at(q).Accuracy << "\t" << listPerformance[p].at(q).GMean
					<< "\t" << listPerformance[p].at(q).nSV << "\n";
			}
		}
		fout.close();
	}

	listPerformance.clear();

	return 0;
}
>

Instances.h

<
#pragma once
#include <vector>
#include <shark/Data/Dataset.h>
#include <shark/Data/Csv.h>
#include <shark/Data/WeightedDataset.h>
using namespace shark;
class Instances
{
   
public:

	LabeledData<RealVector, unsigned int> m_LabeledData;
	WeightedLabeledData<RealVector, unsigned int> m_WeightedLabeledData;

	int m_ClassIndex;
	int m_TotalAttributes;
        int m_numberOfClass;

public:
	Instances()
	{

	}
        
	Instances(const Instances&) = default;				// Copy constructor
	Instances(Instances&&);								// Move constructor
	Instances& operator =(const Instances &) = default;
	std::string filename = "data/training.csv";
	Instances(std::string filename);
	Instances(const LabeledData<RealVector, unsigned int>& labeledData);
	
        //************************************
	// Method:    setClassIndex
	// FullName:  Instances::setClassIndex
	// Access:    public 
	// Returns:   void
	// Qualifier: Convert unlabeled data into labeled data. The count for each class is recorded
	// Parameter: int index
	//************************************
	//void SetClassIndex(int index);

	LabeledData<RealVector, unsigned int> GetLabeledData();
	//WeightedLabeledData<RealVector, unsigned int> GetWeightedLabeledData();

	//************************************
	// Method:    GetTotalAttributes
	// FullName:  Instances::GetTotalAttributes
	// Access:    public 
	// Returns:   int
	// Qualifier: Return total attributes in this data, inclusive class
	//************************************
	size_t GetTotalAttributes();
        size_t GetTotalClasses();
	

	virtual ~Instances();

private:
	//void createData(Data<RealVector>& data);
	//void createData(LabeledData<RealVector, unsigned int>& data);
	//void createData(WeightedLabeledData<RealVector, unsigned int>& data);
};
>

Instances.cpp
<
#include "Instances.h"
Instances::Instances(std::string filename)
{
	//Assuming that the first row is the title
        std::cout << filename;
	importCSV(m_LabeledData, filename,shark::LAST_COLUMN);
	m_LabeledData.makeIndependent();
	//m_TotalAttributes = m_LabeledData.inputs().element(0).size();
	m_TotalAttributes = inputDimension(m_LabeledData);
	//std::cout << m_TotalAttributes;
}

//Instances::Instances(const Instances& instances) : m_UnlabeledData(instances.m_UnlabeledData), m_LabeledData(instances.m_LabeledData)
//{
//
//}

Instances::Instances(Instances&& instances)
{
	m_LabeledData = instances.m_LabeledData;
	m_LabeledData.makeIndependent();
}

Instances::Instances(const LabeledData<RealVector, unsigned int>& labeledData)
{
	m_LabeledData = labeledData;
	m_LabeledData.makeIndependent();
	m_TotalAttributes = inputDimension(m_LabeledData);
}

LabeledData<RealVector, unsigned int> Instances::GetLabeledData()
{
	return m_LabeledData;
}

//WeightedLabeledData<RealVector, unsigned int> Instances::GetWeightedLabeledData()
//{
//	return m_WeightedLabeledData;
//}

size_t Instances::GetTotalAttributes()
{
	return m_TotalAttributes;
}

Instances::~Instances()
{

}
>

**Comment by Ulfgard** [comment]
Hi. This code does not really minimal. The error is probably within the file you want to read in. E.g the format does not fit to the chosen parameters.

**Comment by amaad** [comment]
i am working on a project just download a code. code requires 11 command line argument but i am stuck in this loop cannot read the input file here  

for (int i = 0; i < 10; ++i)
		{
                        cout << trainFiles[t* 10 + i] <<endl;
                      
                            Instances trainInstances(trainFiles[t* 10 + i]);
			Instances testInstances(testFiles[t* 10 + i]);

**Comment by Ulfgard** [comment]
As I said: your problem is most likely that your data files does not match what importCSV expects based on the arguments you provide:

importCSV(m_LabeledData, filename,shark::LAST_COLUMN);

documentation:
http://image.diku.dk/shark/doxygen_pages/html/group__shark__globals.html#ga4d912e42d355b5bb83c25a6967795d74

especially it expects that the numbers are separated by a comma and that comments start with a #.

**Comment by amaad** [comment]
now i am facing this error 

terminate called after throwing an instance of 'shark::Exception'
  what():  [/home/amaad/opt/Shark/include/shark/Data/Csv.h::importCSV,412] Stream cannot be opened for reading.

**Comment by Ulfgard** [comment]
wrong filename.
________________________________
